### PR TITLE
[10.x] Improve decimal shape validation

### PR DESCRIPTION
**Pull request by timacdonald**
The decimal rules checks that a value is numeric and then checks the value against a regex.

The current decimal validation regex is expecting:

`[+-]?` an optional positive / negative symbol

`\d*` zero or more digits

`.` any character, except new lines

`\d*` zero of more digits

All of the following values would pass this regex:

```
+
-
123
+123
-123
123@
123.
+123@
-123@
+123.
-123.
123@456
123.456
+123@456
-123@456
+123.456
-123.456
```

However, we first check that the value is `numeric` in PHP's eyes, so that reduces the values that get checked against the regex to the following set:

```
123
+123
-123
123.
+123.
-123.
123.456
+123.456
-123.456
```

but scientific notation will also pass the `is_numeric` check. So we add the following to the possible values. You will note, however, that these do not match the regex:

```
1.2e3
7E-10
```

When the `preg_match` happens, it has `0` results in `$matches`, and the rule interprets this as the number having `0` decimal places.

I believe this is a bug and the decimal rule should in fact check that the string matches the expected pattern, i.e., the decimal rule fails for scientific notation.

### CODE REVIEW

File: src/Illuminate/Validation/Concerns/ValidatesAttributes.php [hunk 1]
```diff
@@ -599,7 +599,9 @@ public function validateDecimal($attribute, $value, $parameters)
 
         $matches = [];
 
-        preg_match('/^[+-]?\d*.(\d*)$/', $value, $matches);
+        if (preg_match('/^[+-]?\d*\.?(\d*)$/', $value, $matches) !== 1) {
+            return false;
+        }
```
Comment (timacdonald) [inline src/Illuminate/Validation/Concerns/ValidatesAttributes.php:604]: The "dot" has been escaped here. I believe this was a mistake in the original PR.

The dot allows 1 of _any_ character. I believe it should only be a literal `.`.

**Comment by vlakoff** [comment]
Nice, you have fixed the concerns I had in https://github.com/laravel/framework/pull/45356#issuecomment-1365524941 and following, which didn't get interest then…

With your code:
* pass: `42` / `.42` / `42.` – fine
* doesn't pass: ` .42` / `.42 ` (leading/trailing spaces) – fine too

I noticed a caveat though: it doesn't make sure there is at least one digit,
so the followings would erroneously pass: `.` / `+.` / `-.`

**Comment by vlakoff** [comment]
Empty strings would also pass the regex.

BUT, there is the `validateNumeric` before the regex. Thanks to this, `.` / `+.` / `-.` / `<empty>` get rejected actually.

**Comment by vlakoff** [comment]
After trying various regex features, I figured out a nice way to validate everything without relying on a `validateNumeric` before.

Let's make use of [branch reset](https://www.rexegg.com/regex-disambiguation.php#branchreset):

```php
if (preg_match('/^[+-]?(?|\d+(\.\d*)?|(\.\d+))$/', $value, $matches) !== 1) {
    return false;
}

$decimals = isset($matches[1]) ? (strlen($matches[1]) - 1) : 0;
```

* pass: `42` / `42.` / `42.3` / `.3`, with optional `+` or `-`
* don't pass: `.` / `+` / `-` / `+.` / `-.` / empty string

**Comment by vlakoff** [comment]
Also, leading/trailing spaces should be supported, for consistency with [`validateNumeric`](https://github.com/laravel/framework/blob/9d7c869cd53c5c18b1939a7d826506f31f74669c/src/Illuminate/Validation/Concerns/ValidatesAttributes.php#L1786) and [`validateMultipleOf`](https://github.com/laravel/framework/blob/9d7c869cd53c5c18b1939a7d826506f31f74669c/src/Illuminate/Validation/Concerns/ValidatesAttributes.php#L1724) methods. We simply have to add `\s*` at begin and end of the regex.

**Pull request merged**